cmd/scollector/collectors/snmp.go

```diff
@@ -67,3 +91,62 @@ func snmp_oid(host, community, oid string) (*big.Int, error) {
 	err := snmp.Get(host, community, oid, &v)
 	return v, err
 }
+
+func c_snmp_generic(cfg conf.SNMP, mib conf.MIB) (opentsdb.MultiDataPoint, error) {
+	md := opentsdb.MultiDataPoint{}
+	baseOid := mib.BaseOid
+
+	for _, metric := range mib.Metrics {
+		rate := metadata.RateType(metric.RateType)
+		if rate == "" {
+			rate = metadata.Gauge
+		}
+		unit := metadata.Unit(metric.Unit)
+		if unit == "" {
+			unit = metadata.None
+		}
+		tagset := opentsdb.TagSet{"host": cfg.Host}
+		v, err := snmp_oid(cfg.Host, cfg.Community, baseOid+metric.Oid)
+		if err != nil && metric.FallbackOid != "" {
+			v, err = snmp_oid(cfg.Host, cfg.Community, baseOid+metric.FallbackOid)
+		}
+		if err != nil {
+			return md, err
+		}
+		Add(&md, metric.Metric, v, tagset, rate, unit, metric.Description)
+	}
+	for _, tree := range mib.Trees {
+		treeOid := baseOid + tree.BaseOid
+		names, err := snmp_subtree(cfg.Host, cfg.Community, treeOid+tree.LabelSourceOid)
+		if err != nil {
+			return md, err
+		}
+		tagset := opentsdb.TagSet{"host": cfg.Host}
+		for _, metric := range tree.Metrics {
+			rate := metadata.RateType(metric.RateType)
+			if rate == "" {
+				rate = metadata.Gauge
+			}
+			unit := metadata.Unit(metric.Unit)
+			if unit == "" {
+				unit = metadata.None
+			}
+			nodes, err := snmp_subtree(cfg.Host, cfg.Community, treeOid+metric.Oid)
+			if err != nil {
+				return md, err
+			}
+			if len(nodes) != len(names) {
+				return md, fmt.Errorf("snmp tree for %s, and names have different lengths.", metric.Metric)
+			}
+			for i, v := range nodes {
+				tagVal, ok := names[i]
+				if !ok {
+					return md, fmt.Errorf("Tree for tag %s has no entry for metric %s index %d.", tree.TagKey, metric.Metric, i)
+				}
+				tagset[tree.TagKey] = fmt.Sprintf("%s", tagVal)
```

Comment:
If two different metrics have different tree.TagKey values, then the first metric's tagkey will still be in this map during the second metric's loop. Need to recreate this map each metric.